### PR TITLE
feat(security): add GraySwan Cygnal security analyzer

### LINKED ISSUE #1951: incorporate cygnal based security analyzer
We have security analyzers here: https://github.com/OpenHands/software-agent-sdk/tree/30e22095607254194ab1f4d8859f658abe5628a1/openhands-sdk/openhands/sdk/security

In the previous OpenHands/OpenHands repo we have an example of how to use this, but it is not incorporated in this repo. port it over to this repo, and then I'll get you an API key to try it.

### PR DESCRIPTION
**Pull request by neubig**
## Summary

This PR adds a new security analyzer that integrates with GraySwan's Cygnal API for AI safety monitoring. The analyzer evaluates agent actions and conversation history to detect potential security risks.

Fixes #1951

### Key features:
- `GraySwanAnalyzer` class that implements `SecurityAnalyzerBase`
- Converts SDK events to OpenAI message format for API calls
- Maps violation scores to `SecurityRisk` levels (LOW, MEDIUM, HIGH)
- Automatic escalation to HIGH risk for indirect prompt injection (IPI)
- Configurable thresholds with validation (low_threshold < medium_threshold)
- Uses `SecretStr` for API key to prevent accidental exposure in logs
- Uses `httpx` for HTTP requests (already a dependency)

### Environment variables:
- `GRAYSWAN_API_KEY`: Required API key for authentication
- `GRAYSWAN_POLICY_ID`: Optional custom policy ID (defaults to GraySwan's coding agent policy)

### Usage example:
```python
from pydantic import SecretStr
from openhands.sdk.security import GraySwanAnalyzer

# Initialize with API key from environment
analyzer = GraySwanAnalyzer()

# Or pass API key directly (using SecretStr for security)
analyzer = GraySwanAnalyzer(api_key=SecretStr("your-api-key"))

# Set conversation history for context
analyzer.set_events(conversation_events)

# Analyze an action
risk = analyzer.security_risk(action_event)
```

## Checklist

- [x] If the PR is changing/adding functionality, are there tests to reflect this?
- [x] If there is an example, have you run the example to make sure that it works?
- [x] If there are instructions on how to run the code, have you followed the instructions and made sure that it works?
- [ ] If the feature is significant enough to require documentation, is there a PR open on the OpenHands/docs repository with the same branch name?
- [x] Is the github CI passing?

@neubig can click here to [continue refining the PR](https://app.all-hands.dev/conversations/0094a8be44f9407eb5dabf2d2e406cd8)

---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/OpenHands/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Architectures | Base Image | Docs / Tags |
|---|---|---|---|
| java | amd64, arm64 | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | amd64, arm64 | `nikolaik/python-nodejs:python3.12-nodejs22` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.12-nodejs22) |
| golang | amd64, arm64 | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |

**Pull (multi-arch manifest)**
```bash
# Each variant is a multi-arch manifest supporting both amd64 and arm64
docker pull ghcr.io/openhands/agent-server:86ced1b-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-86ced1b-python \
  ghcr.io/openhands/agent-server:86ced1b-python
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-server:86ced1b-golang-amd64
ghcr.io/openhands/agent-server:86ced1b-golang_tag_1.21-bookworm-amd64
ghcr.io/openhands/agent-server:86ced1b-golang-arm64
ghcr.io/openhands/agent-server:86ced1b-golang_tag_1.21-bookworm-arm64
ghcr.io/openhands/agent-server:86ced1b-java-amd64
ghcr.io/openhands/agent-server:86ced1b-eclipse-temurin_tag_17-jdk-amd64
ghcr.io/openhands/agent-server:86ced1b-java-arm64
ghcr.io/openhands/agent-server:86ced1b-eclipse-temurin_tag_17-jdk-arm64
ghcr.io/openhands/agent-server:86ced1b-python-amd64
ghcr.io/openhands/agent-server:86ced1b-nikolaik_s_python-nodejs_tag_python3.12-nodejs22-amd64
ghcr.io/openhands/agent-server:86ced1b-python-arm64
ghcr.io/openhands/agent-server:86ced1b-nikolaik_s_python-nodejs_tag_python3.12-nodejs22-arm64
ghcr.io/openhands/agent-server:86ced1b-golang
ghcr.io/openhands/agent-server:86ced1b-java
ghcr.io/openhands/agent-server:86ced1b-python
```

**About Multi-Architecture Support**
- Each variant tag (e.g., `86ced1b-python`) is a **multi-arch manifest** supporting both **amd64** and **arm64**
- Docker automatically pulls the correct architecture for your platform
- Individual architecture tags (e.g., `86ced1b-python-amd64`) are also available if needed

### CODE REVIEW

**Comment by neubig** [comment]
@openhands check the comments, reflect the ones that seem relevant, and use the API to mark the resolved ones as "resolved". monitor ci and make sure it passes

**Comment by neubig** [comment]
https://github.com/OpenHands check the comments, reflect the ones that seem relevant, and use the API to mark the resolved ones as "resolved". monitor ci and make sure it passes

File: openhands-sdk/openhands/sdk/security/grayswan/analyzer.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        default=0.3,
+        description="Risk threshold for LOW classification",
+    )
+    medium_threshold: float = Field(
+        default=0.7,
+        description="Risk threshold for MEDIUM classification",
+    )
+    high_threshold: float = Field(
+        default=1.0,
+        description="Risk threshold for HIGH classification",
```
[the remark this answers is not in the document]
Comment (neubig) [reply openhands-sdk/openhands/sdk/security/grayswan/analyzer.py]: Fixed: Removed the unused `high_threshold` parameter. The mapping now only uses `low_threshold` and `medium_threshold` as documented.

File: openhands-sdk/openhands/sdk/security/grayswan/analyzer.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    api_url: str = Field(
+        default="https://api.grayswan.ai/cygnal/monitor",
+        description="GraySwan Cygnal API endpoint",
+    )
+    api_key: str | None = Field(
+        default=None,
+        description="GraySwan API key (via GRAYSWAN_API_KEY env var)",
+    )
+    policy_id: str | None = Field(
+        default=None,
```
[the remark this answers is not in the document]
Comment (neubig) [reply openhands-sdk/openhands/sdk/security/grayswan/analyzer.py:74]: Fixed: Changed `api_key` to use `SecretStr` from pydantic. Updated all references to use `.get_secret_value()` when accessing the actual key value.

File: openhands-sdk/openhands/sdk/security/grayswan/analyzer.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+        self._events = list(events)
+
+    def _get_client(self) -> httpx.Client:
+        """Get or create HTTP client."""
+        if self._client is None or self._client.is_closed:
+            self._client = httpx.Client(
+                timeout=self.timeout,
+                headers={
+                    "Authorization": f"Bearer {self.api_key}",
+                    "x-grayswan-api-key": self.api_key or "",
```
[the remark this answers is not in the document]
Comment (neubig) [reply openhands-sdk/openhands/sdk/security/grayswan/analyzer.py]: Fixed: Removed the redundant `x-grayswan-api-key` header. The GraySwan API accepts authentication via Bearer token only.

File: openhands-sdk/openhands/sdk/security/grayswan/analyzer.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+            payload = {"messages": messages, "policy_id": self.policy_id}
+
+            logger.debug(
+                f"Sending request to GraySwan API with {len(messages)} messages "
+                f"and policy_id: {self.policy_id}"
+            )
+
+            response = client.post(self.api_url, json=payload)
+
```
[the remark this answers is not in the document]
Comment (neubig) [reply openhands-sdk/openhands/sdk/security/grayswan/analyzer.py:196]: Fixed: Added validation for missing `violation` field. Now returns `SecurityRisk.UNKNOWN` and logs an error if the field is missing.

File: openhands-sdk/openhands/sdk/security/grayswan/analyzer.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        default=None,
+        description="GraySwan API key (via GRAYSWAN_API_KEY env var)",
+    )
+    policy_id: str | None = Field(
+        default=None,
+        description="GraySwan policy ID (via GRAYSWAN_POLICY_ID env var)",
+    )
+
+    # Internal state - not serialized (using PrivateAttr for Pydantic)
+    _client: httpx.Client | None = PrivateAttr(default=None)
```
[the remark this answers is not in the document]
Comment (neubig) [reply openhands-sdk/openhands/sdk/security/grayswan/analyzer.py:79]: Fixed: Changed `_events` type to `list[LLMConvertibleEvent]` and updated `set_events` to accept `Sequence[LLMConvertibleEvent]`.

File: openhands-sdk/openhands/sdk/security/grayswan/utils.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+                    "arguments": event.tool_call.arguments,
+                },
+            }
+
+            # Remove security_risk from arguments to avoid biasing the analysis
+            try:
+                args = json.loads(event.tool_call.arguments)
+                if "security_risk" in args:
+                    del args["security_risk"]
+                tool_call_dict["function"]["arguments"] = json.dumps(args)
```
[the remark this answers is not in the document]
Comment (neubig) [reply openhands-sdk/openhands/sdk/security/grayswan/utils.py]: Fixed: Added debug logging for JSON parsing failures when removing security_risk from arguments.

File: openhands-sdk/openhands/sdk/security/grayswan/utils.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            content_parts = []
+            for content in llm_message.content:
+                if isinstance(content, TextContent):
+                    content_parts.append(content.text)
+                elif isinstance(content, ImageContent):
+                    # Skip images for security analysis
+                    continue
+
+            content_str = " ".join(content_parts)
+
```
[the remark this answers is not in the document]
Comment (neubig) [reply openhands-sdk/openhands/sdk/security/grayswan/utils.py:70]: Fixed: Added debug logging when images are skipped in security analysis.

File: tests/sdk/security/grayswan/test_grayswan_analyzer.py [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+    def test_map_medium_violation(self, analyzer: GraySwanAnalyzer):
+        """Test that medium violation scores map to MEDIUM risk."""
+        assert analyzer._map_violation_to_risk(0.31) == SecurityRisk.MEDIUM
+        assert analyzer._map_violation_to_risk(0.5) == SecurityRisk.MEDIUM
+        assert analyzer._map_violation_to_risk(0.7) == SecurityRisk.MEDIUM
+
+    def test_map_high_violation(self, analyzer: GraySwanAnalyzer):
+        """Test that high violation scores map to HIGH risk."""
+        assert analyzer._map_violation_to_risk(0.71) == SecurityRisk.HIGH
+        assert analyzer._map_violation_to_risk(0.9) == SecurityRisk.HIGH
```
[the remark this answers is not in the document]
Comment (neubig) [reply tests/sdk/security/grayswan/test_grayswan_analyzer.py:165]: Fixed: Added boundary condition tests for both low and medium thresholds.

**Comment by neubig** [comment]
@openhands fix failing CI, including reflecting any relevant review comments, closing the unclosed comments using graphql, and changing the PR description to match appropriately. I will also provide you an API key that you can use to test the code live, so make sure that it actually works.

**Comment by neubig** [comment]
@openhands reflect the review comments, resolve them using the graphql API, and make sure that CI passes

**Comment by neubig** [comment]
@openhands the sdk-tests github action is stalling forever, probably due to a lock introduced somewhere in this PR. think carefully about all the code that changed and 5 reasons why this might have possibly happened. try to do a minimal reproduction to try to reproduce the issue locally, and then push any necessary changes back and make sure that CI passes.

**Comment by neubig** [comment]
@openhands, this PR is changing more things than it should. check the change and make sure that it is minimal, pre-commit passes, and tests pass.
